src/dfx/src/lib/operations/canister_migration.rs

```diff
@@ -0,0 +1,169 @@
+    SameSubnet(Reserved),
+    CallerNotController { canister: Principal },
+    NotController { canister: Principal },
+    MigratedNotStopped(Reserved),
+    MigratedNotReady(Reserved),
+    ReplacedNotStopped(Reserved),
+    ReplacedHasSnapshots(Reserved),
+    MigratedInsufficientCycles(Reserved),
+    CallFailed { reason: String },
+}
+
+impl fmt::Display for ValidationError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ValidationError::MigrationsDisabled(Reserved) => {
+                write!(f, "Canister migrations are disabled at the moment.")
+            }
+            ValidationError::RateLimited(Reserved) => write!(
+                f,
+                "Canister migration has been rate-limited. Try again later."
+            ),
+            ValidationError::ValidationInProgress { canister } => write!(
+                f,
+                "Validation for canister {canister} is already in progress."
+            ),
+            ValidationError::MigrationInProgress { canister } => write!(
+                f,
+                "Canister migration for canister {canister} is already in progress."
+            ),
+            ValidationError::CanisterNotFound { canister } => {
+                write!(f, "The canister {canister} does not exist.")
+            }
+            ValidationError::SameSubnet(Reserved) => {
+                write!(f, "Both canisters are on the same subnet.")
+            }
+            ValidationError::CallerNotController { canister } => write!(
+                f,
+                "The canister {canister} is not controlled by the calling identity."
+            ),
+            ValidationError::NotController { canister } => write!(
+                f,
+                "The NNS canister sbzkb-zqaaa-aaaaa-aaaiq-cai is not a controller of canister {canister}."
+            ),
+            ValidationError::MigratedNotStopped(Reserved) => {
+                write!(f, "The migrated canister is not stopped.")
+            }
+            ValidationError::MigratedNotReady(Reserved) => write!(
+                f,
+                "The migrated canister is not ready for migration. Try again later."
+            ),
+            ValidationError::ReplacedNotStopped(Reserved) => {
+                write!(f, "The replaced canister is not stopped.")
+            }
+            ValidationError::ReplacedHasSnapshots(Reserved) => {
+                write!(f, "The replaced canister has snapshots.")
+            }
+            ValidationError::MigratedInsufficientCycles(Reserved) => write!(
+                f,
+                "The migrated canister does not have enough cycles for canister migration. Top up the migrated canister with the required amount of cycles."
+            ),
+            ValidationError::CallFailed { reason } => {
+                write!(f, "Internal IC error: a call failed due to {reason}")
+            }
+        }
+    }
+}
```

Comment:
We usually do this with `thiserror`, e.g. like here: https://github.com/dfinity/sdk/blob/master/src/canisters/frontend/ic-asset/src/error/create_chunk.rs